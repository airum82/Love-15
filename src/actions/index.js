export const fetchCourts = (nearbyCourts) => ({
  type: 'COURTS',
  nearbyCourts
})

export const createAccount = (accountInfo) => ({
  type: 'CREATE_ACCOUNT',
  accountInfo
})

export const logIn = (accountInfo) => ({
  type: 'LOG_IN',
  accountInfo
})

export const logOut = () => ({
  type: 'LOG_OUT',
})

export const fetchAccount = (accountInfo) => ({
  type: 'FETCH_ACCOUNT',
  email: accountInfo
})