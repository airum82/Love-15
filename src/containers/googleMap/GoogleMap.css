.map-container {
  text-align: center;
  margin-top: 5%;
}
.map-container input {
  font-size: 1.5em;
}

.map-container h5 {
  font-size: 1.3em;
  color: #4c4c48;
}

.map-container form {
  display: flex;
  justify-content: center;
  margin: 0 auto;
}

.map-container button {
  padding: 1%;
  background-color: #60b0f4;
  border-radius: 0 22px 22px 0;
  color: white;
  font-size: 1.2em;
  border: none;
}