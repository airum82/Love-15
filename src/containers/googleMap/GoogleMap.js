import { GoogleApiWrapper, Map, Marker } from 'google-maps-react';
import React, { Component } from 'react';
import { mapsKey, geoKey } from '../../APIkey';
import { cleanPlaces } from '../../Cleaner/cleaner';
import { connect } from 'react-redux';
import { fetchCourts } from '../../actions';
import './GoogleMap.css'

export class MapContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      coords: {},
      location: '',
      map: '',
      errorMessage: ''
    }
  }

  fetchCoords = async (location) => {
    const splitLocation = location.split(' ');
    const url = 'https://maps.googleapis.com/maps/api/geocode/json?address='
      + splitLocation[0] + ',+' + splitLocation[1] + '&key=' + geoKey;
    try {
      const response = await fetch(url);
      const result = await response.json();
      this.setState({
      coords: result.results[0].geometry.location
      })
    } catch(error) {
        this.setState({
          errorMessage: error.message
        })
    }
  }

  fetchPlaces = (mapProps, map) => {
    const coords = this.state.coords;
    const { google } = mapProps;
    const service = new google.maps.places.PlacesService(map);
    service.nearbySearch({
      location: { lat: coords.lat, lng: coords.lng },
      radius: 17000,
      keyword: ['tennis']
    }, (result) => {
      this.setState({
        places: cleanPlaces(result)
      })
      this.props.handlefetchCourts(cleanPlaces(result));
    })
  };

  handleLocationEntry = (e) => {
    this.setState({
      location: e.target.value
    })
  }

  makeMap = () => {
    this.setState({
      map: ''
    })
    this.setState({
      map: <Map
        className="map"
        onReady={this.fetchPlaces}
        google={this.props.google}
        zoom={14}
        visible={false}
        style={{
          width: '50%',
          height: '50vh'
        }}
        initialCenter={this.state.coords}
        center={this.state.coords}
      />
    })
  }

  render() {
    return (
      <div className="map-container">
        <form onSubmit={async (e) => {
          e.preventDefault();
          await this.fetchCoords(this.state.location);
          this.makeMap();
        }}>
          <input type='text' onChange={this.handleLocationEntry}/>
          <button>submit</button>
        </form>
        {this.state.map}
      </div>
    )
  }
 }

 export const mapStateToProps = (state) => ({
   closeCourts: state.closeCourts
 });

 export const mapDispatchToProps = (dispatch) => ({
  handlefetchCourts: (courts) => dispatch(fetchCourts(courts))
 });

export default connect(mapStateToProps, mapDispatchToProps)(GoogleApiWrapper({
  apiKey: mapsKey
})(MapContainer));