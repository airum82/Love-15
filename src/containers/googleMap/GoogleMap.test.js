import React, { Component } from 'react';
import { shallow, mount } from 'enzyme';
import { MapContainer, mapStateToProps, mapDispatchToProps } from './GoogleMap';
import { geoKey } from '../../APIKey';
import { fetchCourts } from '../../actions';

describe('MapContainer', () => {

  it('should match snapshot upon initial render', () => {
    const wrapper = shallow(<MapContainer />);

    expect(wrapper).toMatchSnapshot();
  })

  it('should call fetchcoords upon submit', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({
        results: [
          {
            geometry: {
              location: {}
            }
          }
        ]
      })
    }))
    const wrapper = mount(<MapContainer />);
    const spy = jest.spyOn(wrapper.instance(), 'fetchCoords');
    wrapper.find('form').simulate('submit');
    expect(spy).toHaveBeenCalled();
  })

  it('fetchCoords should call fetch with correct params', () => {
    const url = 'https://maps.googleapis.com/maps/api/geocode/json?address=denver,+co&key=' + geoKey;
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({
        results: [
          { geometry: {
              location: {}
            } 
          }
        ]
      })
    }))
    const wrapper = shallow(<MapContainer />);
    wrapper.instance().fetchCoords('denver co');
    expect(window.fetch).toHaveBeenCalledWith(url);
  })

  it('fetchCoords should setState with coordinates of location', async () => {
    const url = 'https://maps.googleapis.com/maps/api/geocode/json?address=denver,+co&key=' + geoKey;
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({
        results: [
          {
            geometry: {
              location: {
                lat: 15,
                lng: 12
              }
            }
          }
        ]
      })
    }))
    const initialState = {
      coords: {},
      location: '',
      map: '',
      errorMessage: ''
    }
    const expectedState = {
      coords: {
        lat: 15,
        lng: 12
      },
      location: '',
      map: '',
      errorMessage: ''
    }
    const wrapper = shallow(<MapContainer />);
    expect(wrapper.state()).toEqual(initialState);
    await wrapper.instance().fetchCoords('denver co');
    expect(wrapper.state()).toEqual(expectedState);
  })

  it('should set state with error message if fetch fails', async () => {
    const url = 'https://maps.googleapis.com/maps/api/geocode/json?address=denver,+co&key=' + geoKey;
    window.fetch = jest.fn().mockImplementation(() => Promise.reject(
      new Error('error, invalid location')
    ))
    const initialState = {
      coords: {},
      location: '',
      map: '',
      errorMessage: ''
    }
    const expectedState = {
      coords: {},
      location: '',
      map: '',
      errorMessage: 'error, invalid location'
    }
    const wrapper = shallow(<MapContainer />);
    expect(wrapper.state()).toEqual(initialState);
    await wrapper.instance().fetchCoords('denver co');
    wrapper.update();
    expect(wrapper.state()).toEqual(expectedState);
  })

  it('mapStateToProps should return an object with a closeCourts key', () => {
    const mockState = {
      closeCourts: [{}, {}]
    }
    const expectedOutput = {
      closeCourts: [{}, {}]
    }
    expect(mapStateToProps(mockState)).toEqual(expectedOutput)
  })

  it('handlefetchCourts should call mockDispatch with correct params', () => {
    const mockDispatch = jest.fn();
    const mappedProps = mapDispatchToProps(mockDispatch);
    const courts = [{}, {}];
    mappedProps.handlefetchCourts(courts);
    expect(mockDispatch).toHaveBeenCalledWith(fetchCourts(courts));
  })

  it('handleLocationEntry should set state on change', () => {
    const wrapper = shallow(<MapContainer />);
    expect(wrapper.state().location).toEqual('');
    const mockEvent = {
      target: {
        value: 'denver'
      }
    }
    wrapper.instance().handleLocationEntry(mockEvent);
    expect(wrapper.state().location).toEqual('denver');
  })
})