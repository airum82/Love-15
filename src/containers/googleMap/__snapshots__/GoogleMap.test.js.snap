// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapContainer should match snapshot upon initial render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MapContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
>
          <input
                    onChange={[Function]}
                    type="text"
          />
          <button>
                    submit
          </button>
</form>,
        "",
      ],
      "className": "map-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onChange={[Function]}
              type="text"
/>,
            <button>
              submit
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "submit",
            },
            "ref": null,
            "rendered": "submit",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
>
            <input
                        onChange={[Function]}
                        type="text"
            />
            <button>
                        submit
            </button>
</form>,
          "",
        ],
        "className": "map-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onChange={[Function]}
                type="text"
/>,
              <button>
                submit
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "submit",
              },
              "ref": null,
              "rendered": "submit",
              "type": "button",
            },
          ],
          "type": "form",
        },
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
